### PR TITLE
tests(viewer): add pptr test for viewer

### PR DESCRIPTION
**Pull request by brendankenny**
As we start major work again on the report, it'd be good to have an easy way to verify we aren't breaking viewer. While viewer does already have compile and (a few) basic unit tests, this PR rips off @wardpeet's hard work from #4640 to make a pptr integration test for viewer as well.

Ideally we'd extract some sort of "use puppeteer to test a live report interface" test and have each individual directory file only in charge of setup and teardown, but I think(?) this is sufficient for now as we have more pressing matters.

### CODE REVIEW

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  const selectors = {
+    audits: '.lh-audit,.lh-timeline-metric,.lh-perf-hint',
+    titles: '.lh-score__title, .lh-perf-hint__title, .lh-timeline-metric__title',
+  };
+
+  it('should load with no errors', async () => {
+    assert.deepStrictEqual(pageErrors, []);
+  });
+
+  it('should contain all categories', async () => {
```
Comment (brendankenny) [inline lighthouse-viewer/test/viewer-test-pptr.js]: below is all basically identical to the extension test

File: lighthouse-cli/test/fixtures/static-server.js [hunk 2]
```diff
@@ -24,16 +24,22 @@ function requestHandler(request, response) {
...
     absoluteFilePath = path.join(__dirname, `${zonePath}/dist/zone.js`);
-  }
-
-  // Disallow file requests outside of LH folder
-  if (!path.parse(absoluteFilePath).dir.startsWith(lhRootDirPath)) {
-    return readFileCallback(new Error('Disallowed path'));
+  } else {
+    // Otherwise, disallow file requests outside of LH folder or from node_modules
+    const filePathDir = path.parse(absoluteFilePath).dir;
+    if (!filePathDir.startsWith(lhRootDirPath) || filePathDir.includes('node_modules')) {
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/static-server.js]: drive-by tightening :)

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      headless: false,
+      executablePath: process.env.CHROME_PATH,
+      args: [],
+    });
+    viewerPage = await browser.newPage();
+    viewerPage.on('pageerror', pageError => pageErrors.push(pageError));
+    await viewerPage.goto(viewerUrl, {waitUntil: 'networkidle2'});
+    const fileInput = await viewerPage.$('#hidden-file-input');
+    await fileInput.uploadFile(sampleLhr);
+    await viewerPage.waitForSelector('.lh-container');
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: Do you want to set a smaller timeout for this and goto? Default is 30s. If the link or gist fails, we should know about it sooner. Speed up the tests.

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    await Promise.all([
+      new Promise(resolve => server.close(resolve)),
+      browser && browser.close(),
+    ]);
+  });
+
+
+  const selectors = {
+    audits: '.lh-audit,.lh-timeline-metric,.lh-perf-hint',
+    titles: '.lh-score__title, .lh-perf-hint__title, .lh-timeline-metric__title',
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: spaces here too: `'.lh-audit, .lh-timeline-metric, .lh-perf-hint',`

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  };
+
+  it('should load with no errors', async () => {
+    assert.deepStrictEqual(pageErrors, []);
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: lol. just check `pageErrors.length === 0`

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    // eslint-disable-next-line
+    this.timeout(10 * 1000);
+
+    server.listen(portNumber, 'localhost');
+
+    // start puppeteer
+    browser = await puppeteer.launch({
+      headless: false,
+      executablePath: process.env.CHROME_PATH,
+      args: [],
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: kill?

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  before(async function() {
+    // eslint-disable-next-line
+    this.timeout(10 * 1000);
+
+    server.listen(portNumber, 'localhost');
+
+    // start puppeteer
+    browser = await puppeteer.launch({
+      headless: false,
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: why do we need headful?

Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: > why do we need headful?

...we're testing as a user? I don't know, I copied from  #4640 :) Does it really matter in this instance?

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 6 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: done

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: whoops, yes, misread that as 3s and thought it sounded great :)

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: done

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: > just check pageErrors.length === 0

I got the idea from the audit errors check below when I had errors at first. The error diff prints out the errors the page was seeing (instead of `x !== 0`), which is actually pretty nice when things go wrong and you don't have to do an extra debug step to print them

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 7 above)
Comment (ebidel) [reply lighthouse-viewer/test/viewer-test-pptr.js]: Ah yea. That's because extensions dont work in headless. You could use headless here but I'm not sure it really matters.

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 8]
```diff
@@ -47,14 +47,13 @@ describe('Lighthouse Viewer', function() {
     browser = await puppeteer.launch({
       headless: false,
       executablePath: process.env.CHROME_PATH,
-      args: [],
     });
     viewerPage = await browser.newPage();
     viewerPage.on('pageerror', pageError => pageErrors.push(pageError));
-    await viewerPage.goto(viewerUrl, {waitUntil: 'networkidle2'});
+    await viewerPage.goto(viewerUrl, {waitUntil: 'networkidle2', timeout: 3000});
```
Comment (ebidel) [inline lighthouse-viewer/test/viewer-test-pptr.js]: These will throw if they timeout. Do you want to catch those as errors?

Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: > These will throw if they timeout. Do you want to catch those as errors?

I think it's fine? It'll throw and fail in `before`, which will fail the tests.

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+          debugString: el.textContent,
+          title: auditTitle ? auditTitle.textContent : 'Audit title unvailable',
+        };
+      });
+    }
+
+    const auditErrors = await viewerPage.$$eval('.lh-debug', getDebugStrings, selectors);
+    const errors = auditErrors.filter(item => item.debugString.includes('Audit error:'));
+    const unexpectedErrrors = errors.filter(item => {
+      return !item.debugString.includes('Required RobotsTxt gatherer did not run');
```
Comment (paulirish) [inline lighthouse-viewer/test/viewer-test-pptr.js]: can you confirm the savedartifacts are out of date?

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 7 above)
Comment (paulirish) [reply lighthouse-viewer/test/viewer-test-pptr.js]: headless! lets do it. 

used it over there because extensions yes.

File: lighthouse-viewer/test/viewer-test-pptr.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  before(async function() {
+    server.listen(portNumber, 'localhost');
+
+    // start puppeteer
+    browser = await puppeteer.launch({
+      headless: false,
+      executablePath: process.env.CHROME_PATH,
+    });
+    viewerPage = await browser.newPage();
+    viewerPage.on('pageerror', pageError => pageErrors.push(pageError));
```
Comment (paulirish) [inline lighthouse-viewer/test/viewer-test-pptr.js]: good call.

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 9 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: yeah, no `RobotsTxt` artifact in `artifacts.json`

File: lighthouse-viewer/test/viewer-test-pptr.js (hunk 7 above)
Comment (brendankenny) [reply lighthouse-viewer/test/viewer-test-pptr.js]: > headless! lets do it.

done

**Comment by brendankenny** [comment]
travis unhappiness was due to report changes in #5009. extension test got updates for that but this obviously did not :)
